### PR TITLE
Add support for Power Query

### PR DESCRIPTION
**Pull request by peterbud**
Adding syntax highlighting for Power Query.
[Documentation](https://docs.microsoft.com/en-us/powerquery-m/)
[Specification](https://docs.microsoft.com/en-us/powerquery-m/power-query-m-language-specification)

### CODE REVIEW

File: components/prism-powerquery.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        /\bValue\.(?:Add|Compare|Divide|Equals|Metadata|Multiply|NativeQuery|NullableEquals|RemoveMetadata|ReplaceMetadata|Subtract|Type)\b/,
+        /\b(?:Byte|Currency|Decimal|Double|Guid|Int8|Int16|Int32|Int64|Percentage|Single)\.From\b/,
+        /\bJson.FromValue\b/
+    ],
+    'data-type':
+    {
+        pattern: /\b(?:any|anynonnull|binary|date|datetime|datetimezone|duration|function|list|logical|none|number|record|table|text|time|type)\b/,
+        alias: 'variable'
+    },
+    'number': /\b0x[\da-f]+\b|[+-]?\b\d+\.?\d*e?\d*|\B\.\d+\b/i,
```
Comment (RunDevelopment) [inline components/prism-powerquery.js]: The [spec](https://docs.microsoft.com/en-us/powerquery-m/power-query-m-language-specification) has a nice grammar for number on pages 20-21.

```suggestion
    'number': /\b0x[\da-f]+\b|[+-]?(?:\b\d+\.|\B\.)?\d+(?:e[+-]?\d+)?/i,
```

Could you also add the following numbers to the numbers test, please?
```
123
123e45
123e+45
123e-45
123E45
123E+45
123E-45
123.78e-45
.78e-45
```

File: components/prism-powerquery.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    'operator': /[-+*\/&?@^]|<(?:=>?|>)?|>(?:=?)|=(?:>?)/,
```
[the remark this answers is not in the document]
Comment (RunDevelopment) [reply components/prism-powerquery.js]: On page 106, it also lists `..` (for lists) and `...` as operators. What about them?

File: components/prism-powerquery.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+Prism.languages.powerquery = {
+    'comment': {
+        pattern: /(^|[^\\])(?:\/\*[\s\S]*?\*\/|(?:\/\/).*)/,
+        lookbehind: true
+    },
+    'quoted-identifier': {
+        pattern: /#"(?:[^"\r\n]|"")*"(?!")/,
+        greedy: true,
+        alias: 'constant'
```
Comment (RunDevelopment) [inline components/prism-powerquery.js]: Aren't these just variable names? Why the `constant` alias?

File: components/prism-powerquery.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        /\bJoinSide\.(?:Left|Right)\b/,
+        /\bPrecision\.(?:Double|Decimal)\b/,
+        /\bRelativePosition\.(?:From(?:End|Start))\b/,
+        /\bTextEncoding\.(?:Ascii|BigEndianUnicode|Unicode|Utf8|Utf16|Windows)\b/,
+        /\bValue\.(?:As|Is|ReplaceType)\b/,
+        /\b(?:Any|Binary|Date|DateTime|DateTimeZone|Duration|Int8|Int16|Int32|Int64|Function|List|Logical|None|Number|Record|Table|Text|Time)\.Type\b/,
+        /\b(?:true|false|null)\b/
+    ],
+    'keyword': /\b(?:and|as|each|else|error|if|in|is|let|meta|not|nullable|optional|or|otherwise|section|shared|then|try|type)\b|#(?:binary|date|datetime|datetimezone|duration|infinity|nan|sections|shared|table|time)\b/,
+    'function': [
```
Comment (RunDevelopment) [inline components/prism-powerquery.js]: Can't we identify functions using the call syntax? 
Wouldn't the following also work:
```js
{
    pattern: /(^|[^#\w.])(?!\d)[\w.]+(?=\s*\()/,
    lookbehind: true
}
```
Or is there another language feature which also uses parenthesis I didn't consider?

File: components/prism-powerquery.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        /\bGroupKind\.(?:Global|Local)\b/,
+        /\bExtraValues\.(?:List|Ignore|Error)\b/,
+        /\bJoinAlgorithm\.(?:Dynamic|PairwiseHash|SortMerge|LeftHash|RightHash|LeftIndex|RightIndex)\b/,
+        /\bJoinSide\.(?:Left|Right)\b/,
+        /\bPrecision\.(?:Double|Decimal)\b/,
+        /\bRelativePosition\.(?:From(?:End|Start))\b/,
+        /\bTextEncoding\.(?:Ascii|BigEndianUnicode|Unicode|Utf8|Utf16|Windows)\b/,
+        /\bValue\.(?:As|Is|ReplaceType)\b/,
+        /\b(?:Any|Binary|Date|DateTime|DateTimeZone|Duration|Int8|Int16|Int32|Int64|Function|List|Logical|None|Number|Record|Table|Text|Time)\.Type\b/,
+        /\b(?:true|false|null)\b/
```
[the remark this answers is not in the document]
Comment (peterbud) [reply components/prism-powerquery.js]: OK, moved to a separate feature. Added also tests

File: components/prism-powerquery.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        /\bRoundingMode\.(?:AwayFromZero|Down|ToEven|TowardZero|Up)\b/,
+        /\bMissingField\.(?:Error|Ignore|UseNull)\b/,
+        /\bQuoteStyle\.(?:Csv|None)\b/,
+        /\bJoinKind\.(?:Inner|LeftOuter|RightOuter|FullOuter|LeftAnti|RightAnti)\b/,
+        /\bGroupKind\.(?:Global|Local)\b/,
+        /\bExtraValues\.(?:List|Ignore|Error)\b/,
+        /\bJoinAlgorithm\.(?:Dynamic|PairwiseHash|SortMerge|LeftHash|RightHash|LeftIndex|RightIndex)\b/,
+        /\bJoinSide\.(?:Left|Right)\b/,
+        /\bPrecision\.(?:Double|Decimal)\b/,
+        /\bRelativePosition\.(?:From(?:End|Start))\b/,
```
[the remark this answers is not in the document]
Comment (peterbud) [reply components/prism-powerquery.js]: Good catch!

File: components/prism-powerquery.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        /\bText\.(?:At|(?:After|Before)Delimiter|BetweenDelimiters|Clean|Combine|Contains|End(?:sWith)?|From(?:Binary)?|InferNumberType|Insert|Length|Lower|Middle|NewGuid|Pad(?:End|Start)|PositionOf(?:Any)?|Proper|Range|Remove(?:Range)?|Repeat|Replace(?:Range)?|Reverse|Select|Split(?:Any)?|Start(?:sWith)?|To(?:Binary|List)|Trim(?:End|Start)?|Upper)\b/,
+        /\bTime\.(?:EndOfHour|From(?:Text)?|Hour|Minute|Second|StartOfHour|To(?:Record|Text))\b/,
+        /\bType\.(?:AddTableKey|ClosedRecord|Facets|For(?:Function|Record)|Function(?:Parameters|RequiredParameters|Return)|Is(?:Nullable|OpenRecord)?|ListItem|NonNullable|OpenRecord|RecordFields|Replace(?:Facets|TableKeys)|Table(?:Column|Keys|Row|Schema)|Union)\b/,
+        /\bUri\.(?:BuildQueryString|Combine|EscapeDataString|Parts)\b/,
+        /\bValue\.(?:Add|Compare|Divide|Equals|Metadata|Multiply|NativeQuery|NullableEquals|RemoveMetadata|ReplaceMetadata|Subtract|Type)\b/,
+        /\b(?:Byte|Currency|Decimal|Double|Guid|Int8|Int16|Int32|Int64|Percentage|Single)\.From\b/,
+        /\bJson.FromValue\b/
+    ],
+    'data-type':
+    {
```
[the remark this answers is not in the document]
Comment (peterbud) [reply components/prism-powerquery.js]: Done

File: components/prism-powerquery.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (peterbud) [reply components/prism-powerquery.js]: Interestingly, I know `..` but I'm not sure where the tripe dot operator `...` is used in real life. Nevertheless I have added them to the pattern, and added test cases for them as well

File: components/prism-powerquery.js (hunk 3 above)
Comment (peterbud) [reply components/prism-powerquery.js]: OK changed to `variable`

File: components/prism-powerquery.js (hunk 1 above)
Comment (peterbud) [reply components/prism-powerquery.js]: That is much better, I made this change, but still there is one issue: This pattern above is now recognizing the double dot `..` operator. Example:
```
AA = .12,
BB = {1..100},
```
It recognizes 1.**.100** as a number (aka it includes one dot from the double dot operator in the number token.

I have tried to fix this, but I'm not so good with regex. As I see I cannot use `(?<!\.)` - could you maybe suggest a good solution pls?

Comment (peterbud) [reply components/prism-powerquery.js]: And also noticed that this pattern recognizes variable names as numbers if the have a digit in their names, like `MyVariable2`

Comment (RunDevelopment) [reply components/prism-powerquery.js]: My bad. This pattern should work.

```js
{
    pattern: /\b0x[\da-f]+\b|(?:[+-]?(?:\b\d+\.)?\b\d+|[+-]\.\d+|(^|[^.])\B\.\d+)(?:e[+-]?\d+)?\b/,
    lookbehind: true
}
```

(Thank god, rewriting the `(?<!\.)` using a Prism lookbehind wasn't tedious at all... \s
Why did JS only add support for lookbehinds two years ago...)

Comment (peterbud) [reply components/prism-powerquery.js]: Super, thanks!

File: components/prism-powerquery.js (hunk 4 above)
Comment (peterbud) [reply components/prism-powerquery.js]: That simplifies the pattern a lot :-0. Works perfectly, thanks

File: tests/languages/powerquery/number_feature.test [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+123
+123e45
+123e+45
+123e-45
+123E45
+123E+45
+123E-45
+123.78e-45
+.78e-45
+
```
Comment (RunDevelopment) [inline tests/languages/powerquery/number_feature.test]: Could you please add an example for a range like `123..456`. This broke the pattern before, so I think we should include a test case.

Comment (peterbud) [reply tests/languages/powerquery/number_feature.test]: OK, I have added it here to check that it properly recognizes the `..` operator

Comment (peterbud) [reply tests/languages/powerquery/number_feature.test]: Should I simplify the testfile for the functions, maybe?

Comment (RunDevelopment) [reply tests/languages/powerquery/number_feature.test]: Yeah, that'd be nice. Maybe also add a case for user-defined functions.

After that, I think this PR is good to go.

Comment (peterbud) [reply tests/languages/powerquery/number_feature.test]: OK done. Function tests are now very simple, hope this is OK.

**Comment by RunDevelopment** [comment]
Thank you for contributing @peterbud!

**Pull request merged**